File: Data Structures & Algorithms/maximum-product-subarray/submission-0.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int res = nums[0];
        int currMax = 1, currMin = 1;

        for(int num : nums){
            int tmp = currMax * num;
            currMax = max(max(tmp, currMin * num), num);
            currMin = min(min(tmp, currMin * num), num);
            res = max(res, currMax);
        }
        return res;
    }
};

File: Data Structures & Algorithms/maximum-subarray/submission-0.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        runningSum, res = 0, nums[0]

        for num in nums:
            if runningSum < 0:
                runningSum = 0 
            
            runningSum += num
            res = max(res, runningSum)

        return res